openmc/cell.py

```diff
@@ -498,19 +521,22 @@ def create_xml_subelement(self, xml_element):
             # tree. When it reaches a leaf (a Halfspace), it creates a <surface>
             # element for the corresponding surface if none has been created
             # thus far.
-            def create_surface_elements(node, element):
+            def create_surface_elements(node, element, memo=None):
                 if isinstance(node, Halfspace):
-                    path = "./surface[@id='{}']".format(node.surface.id)
-                    if xml_element.find(path) is None:
-                        xml_element.append(node.surface.to_xml_element())
+                    if memo and id(node.surface) in memo:
+                        return
+                    if memo is not None:
+                        memo.add(id(node.surface))
```

Comment:
Don't these conflict with cell IDs in the same memo set?

Reply:
Maybe I'm missing something... they should be unique Python object IDs as opposed to the `Cell.id` property, right?

Another option would be to simply add the object instance itself to the set, which is maybe more clear.

Reply:
Oh, haha, somehow when I skimmed through I was thinking `obj.id` instead of `id(obj)`. You're absolutely right.

Reply:
And yes, I think adding the objects to the set would be cleaner.